edx_oauth_client/pipeline.py

```diff
@@ -1,139 +1,164 @@
+from social_core.pipeline import partial
+from third_party_auth.pipeline import AuthEntryError, is_api, get_complete_url
 
-from social.pipeline import partial
+from openedx.core.djangoapps.user_authn import cookies as user_authn_cookies
+from openedx.core.djangoapps.user_authn.views.registration_form import AccountCreationForm
+from openedx.core.djangoapps.user_authn.utils import generate_password
+from student.helpers import do_create_account
 
-from student.views import create_account_with_params, reactivation_email_for_user
-from student.models import UserProfile
-from third_party_auth.pipeline import AuthEntryError, make_random_password
 
 log = logging.getLogger(__name__)
 
-# The following are various possible values for the AUTH_ENTRY_KEY.
-AUTH_ENTRY_LOGIN = 'login'
-AUTH_ENTRY_REGISTER = 'register'
-AUTH_ENTRY_ACCOUNT_SETTINGS = 'account_settings'
-
-AUTH_ENTRY_LOGIN_2 = 'account_login'
-AUTH_ENTRY_REGISTER_2 = 'account_register'
-
-# Entry modes into the authentication process by a remote API call (as opposed to a browser session).
-AUTH_ENTRY_LOGIN_API = 'login_api'
-AUTH_ENTRY_REGISTER_API = 'register_api'
-
 
 @partial.partial
 def ensure_user_information(
-        strategy, auth_entry, backend=None, user=None, social=None,
-        allow_inactive_user=False, *args, **kwargs):
+        strategy, auth_entry, backend=None, user=None, social=None, allow_inactive_user=False, *args, **kwargs
+):
     """
-    Ensure that we have the necessary information about a user (either an
-    existing account or registration data) to proceed with the pipeline.
+    Ensure that we have the necessary information about a user to proceed with the pipeline.
+
+    Either an existing account or registration data.
     """
 
-    response = {}
     data = {}
-
     try:
         if 'data' in kwargs['response']:
             user_data = kwargs['response']['data'][0]
         else:
             user_data = kwargs['response']
+        log.info('Get user data')
 
-        log.info('Get user data: %s', str(user_data))
         data['access_token'] = kwargs['response']['access_token']
 
+        country = user_data.get('country', 'UA')
+
+        if not country:
+            log.info('No country in response.')
+
         for key, value in backend.setting('USER_DATA_KEY_VALUES').items():
             data[key] = user_data.get(value)
 
-        if 'username' not in data or not data['username']:
-            data['username'] = data.get('name', slugify(data['email']))
+        if not data['email']:
+            data['email'] = strategy.session_get('email', "")
 
-        if any((data['first_name'], data['last_name'])):
-            data['name'] = u'{} {}'.format(data['first_name'], data['last_name']).strip()
-        else:
-            data['name'] = user_data.get('preferred_username')
+        data['username'] = data['email']
 
-        if not all((data['username'], data['email'])):
+        if kwargs.get('is_new') and not all((data['username'], data['email'])):
```

Comment:
Do we need to check `not all((data['username'], data['email']))`? We defined them previously

Reply:
I think would be better to check here and rise exception if we didn't receive necessary info.